R/snapshot.R

```diff
@@ -948,39 +957,99 @@ renv_snapshot_report_actions <- function(actions, old, new) {
 }
 # nocov end
 
+# compute the package dependencies inferred for a project,
+# respecting the snapshot type selected (or currently configured)
+# for the associated project
 renv_snapshot_dependencies <- function(project, type = NULL, dev = FALSE) {
 
-  if (!is.null(the$init_dependencies)) {
-    return(the$init_dependencies$Package)
-  }
-
   type <- type %||% settings$snapshot.type(project = project)
 
-  if (type %in% "all")
-    return(installed_packages(field = "Package"))
+  dynamic(
+    list(project = project, type = type, dev = dev),
+    renv_snapshot_dependencies_impl(project, type, dev)
+  )
+
+}
+
+renv_snapshot_dependencies_impl <- function(project, type = NULL, dev = FALSE) {
+
+  if (type %in% "all") {
+    packages <- installed_packages(field = "Package")
+    return(setdiff(packages, renv_packages_base()))
+  }
 
   if (type %in% "custom") {
     filter <- renv_snapshot_filter_custom_resolve()
     return(filter(project))
   }
 
   path <- case(
-    type %in% "implicit" ~ project,
+    type %in% c("packrat", "implicit") ~ project,
     type %in% "explicit" ~ file.path(project, "DESCRIPTION"),
     ~ {
       fmt <- "internal error: unhandled snapshot type '%s' in %s"
       stopf(fmt, type, stringify(sys.call()))
     }
   )
 
-  renv_dependencies_confirm(
-    "snapshot",
-    path = path,
-    root = project,
-    field = "Package",
-    dev = dev
+  packages <- withCallingHandlers(
+
+    renv_dependencies_impl(
+      path = path,
+      root = project,
+      field = "Package",
+      errors = config$dependency.errors(),
+      dev = dev
+    ),
+
+    # require user confirmation to proceed if there's a reported error
+    renv.dependencies.problem = function(cnd) {
+      if (interactive() && !proceed())
+        cancel()
+    }
+
   )
 
+  unique(packages)
+
+}
+
+# compute package records from the provided library paths,
+# normally to be included as part of an renv lockfile
+renv_snapshot_packages <- function(packages, libpaths, project) {
+
+  ignored <- c(
+    renv_packages_base(),
+    renv_project_ignored_packages(project = project),
+    if (is_testing()) "renv"
+  )
+
+  callback <- function(package, location, project) {
+    if (nzchar(location) && !package %in% ignored)
+      return(location)
+  }
+
+  # expand package dependency tree
+  paths <- renv_package_dependencies(
+    packages = packages,
+    libpaths = libpaths,
+    callback = callback,
+    project = project
+  )
+
+  # keep only packages with known locations
+  paths <- convert(filter(paths, is.character), "character")
+
+  # diagnose issues with the scanned packages
+  paths <- uapply(libpaths, function(library) {
+    renv_snapshot_library_diagnose(
+      library = library,
+      paths   = filter(paths, startswith, prefix = library))
+  })
+
+  # now, snapshot the remaining packages
+  records <- map(paths, renv_snapshot_description)
+
 }
 
 renv_snapshot_filter <- function(project, records, type, packages, exclude) {
```

Comment:
Now that we're collecting only the records we know we require (respecting the snapshot type) I wonder if this method is still really needed?